### PR TITLE
Define GetBody of http.Request for automatic retry

### PR DESCRIPTION
**Pull request by tkashem**
**What type of PR is this?**
> /kind api-change

/kind bug

> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Add a unit test that ensures that `rest.Request.body`, if not empty always produces a `http.Request` with a non-nil `GetBody`.

`net/http2` retries a request if it receives a `GOAWAY` frame.  `GOAWAY` frame has `LastStreamID` which represents the last stream identifier the server has processed or is aware of. On the client
side, `net/http2` will set an error (`http2: Transport received Server's graceful shutdown GOAWAY` for any stream that has an identifier higher than that of LastStreamID.

However, if the request body is not empty then net/http2 will retry only when it can reset back the request body to its original state via the optional `req.GetBody`.

This is where it happens in net/http2
https://github.com/golang/go/blob/master/src/net/http/h2_bundle.go#L6991

**Does this PR introduce a user-facing change?**:
NONE

-->
```release-note
NONE
```

### CODE REVIEW

**Comment by sttts** [comment]
Great PR!

Can one test the retry behaviour somehow in a unit test?

**Comment by p0lyn0mial** [comment]
great finding @tkashem

/lgtm

**Comment by tkashem** [comment]
@sttts I am looking into unit test/integration test.

**Comment by fedebongio** [comment]
/cc @caesarxuchao 
(just for fyi)

**Comment by tkashem** [comment]
@sttts after further debugging I found out that net/http sets `GetBody` for some type of request body but not all. I think we can keep it as a safeguard to ensure that all net/http requests with a body have a `GetBody` defined.

Also, as far as unit testing is concerned, there is an inherent race condition between the server sending a `GOAWAY` frame with `LastStreamID=N` and the client generating a new stream `P (P>N)` before it receives the `GOAWAY` frame.
- `E1`: server sends a `GOAWAY` frame `LastStreamID=N`
- `E2`: client receives the `GOAWAY` frame and removes the underlying tcp connection from the pool.

The test needs to generate the new stream between `E1` and `E2`. If the test can manage to generate a new stream `P` between `E1` and `E2` then `P` will be ignored by the server. The client will retry `P` on a new tcp connection and `GetBody` will be utilized by net/http. But we don't have the knobs in net/http/httptest to design such a test in a deterministic way.

If `P` is generated after `E2` then it will be tried on a new tcp connection with a `StreamID=1` and the `GOAWAY` frame sent by the server does not have any impact on `P`.

**Comment by sttts** [comment]
Lgtm. But this needs more eyes as it is super central.

**Comment by sttts** [comment]
/lgtm
/approve
/hold

Hold for more eyes.

File: staging/src/k8s.io/client-go/rest/request.go [hunk 1]
```diff
@@ -869,6 +869,27 @@ func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Resp
...
+		//
+		// This is where it happens in net/http2
+		// https://github.com/golang/go/blob/master/src/net/http/h2_bundle.go#L6991
+		//
+		// Let's set GetBody for our request so net/http2 automatically retry it. It's ideal if
+		// we allow net/http2 to retry a request before we do a retry here.
+		//
+		// net/http sets GetBody for some type of body but not all, so it's good to
+		// ensure that all requests with a body have a GetBody defined.
+		if r.body != nil && req.GetBody == nil {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/request.go:887]: shouldn't this be limited to r.verb == GET or bodies that can be reset? the GetBody implemented here doesn't do anything to reset the body to its original state, which means it is not implemented according to the godoc for GetBody

Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/request.go]: Thanks @liggitt for your feedback. 
The `Body` function sets the `body` of a `rest.Request` object to an instance of `*bytes.Reader` if the input is `Runtime.Object`, `[]byte` or `string`. On the other hand, If body is of type `*bytes.Reader` then `http.NewRequest` will set a `GetBody` for us - https://github.com/golang/go/blob/master/src/net/http/request.go#L881-L887

So I think we are covered, I have reduced my PR to a unit test that ensures that `rest.Request.body`, if not empty always produces a `http.Request` with a non-nil `GetBody`.

**Pull request closed without merging**